Lib/fontTools/ttLib/tables/O_S_2f_2.py

```diff
@@ -475,22 +476,26 @@ def recalcUnicodeRanges(self, ttFont, pruneOnly=False):
 )
 
 
-_unicodeRangeSets = []
+_unicodeSteps = []
+_unicodeValues = []
 
-def _getUnicodeRangeSets():
-	# build the sets of codepoints for each unicode range bit, and cache result
-	if not _unicodeRangeSets:
+def _getUnicodeRanges():
+	# build the ranges of codepoints for each unicode range bit, and cache result
+	if not _unicodeSteps:
 		for bit, blocks in enumerate(OS2_UNICODE_RANGES):
-			rangeset = set()
 			for _, (start, stop) in blocks:
-				rangeset.update(set(range(start, stop+1)))
-			if bit == 57:
-				# The spec says that bit 57 ("Non Plane 0") implies that there's
-				# at least one codepoint beyond the BMP; so I also include all
-				# the non-BMP codepoints here
-				rangeset.update(set(range(0x10000, 0x110000)))
-			_unicodeRangeSets.append(rangeset)
-	return _unicodeRangeSets
+				index = bisect.bisect(_unicodeSteps, start)
+				_unicodeSteps.insert(index, start)
+				_unicodeValues.insert(index, bit)
+				index = bisect.bisect_left(_unicodeSteps, stop + 1)
+				_unicodeSteps.insert(index, stop + 1)
+				_unicodeValues.insert(index, None)
+		for i, (step, value) in enumerate(zip(_unicodeSteps, _unicodeValues)):
+			if i and _unicodeSteps[i] == _unicodeSteps[i-1]:
+				_unicodeSteps.pop(i - 1)
+				_unicodeValues.pop(i - 1)
+		_unicodeValues.insert(0, None)
+	return _unicodeSteps, _unicodeValues
```

Comment:
This seems a bit complicated. The ranges don't overlap, so you could make a list like this:
```python
values = []
for bit, val in enumerate(OS2_UNICODE_RANGES):
    for name, (rng_start, rng_end) in val:
        values.append((rng_start, rng_end, bit))
values.sort()
```
And either do a bisect on that directly, or unpack it in three separate lists, and bisect on the range start values.

But yeah, I was just thinking that bisecting may give the best compromise between loading speed, memory usage and lookup performance, so I'm glad you did this!

Reply:
> This seems a bit complicated. The ranges don't overlap, so you could make a list like this:

Of course, you’re right.

> But yeah, I was just thinking that bisecting may give the best compromise between loading speed, memory usage and lookup performance, so I'm glad you did this!

:smile: